docs/Generators.md

```diff
@@ -370,3 +370,97 @@ Any `config.json` files found under the `cluster-config` directory will be param
     Only JSON file parsing is currently supported. The work to add support for YAML files is [tracked here](https://github.com/argoproj-labs/applicationset/issues/106).
 
 As with other generators, clusters *must* already be defined within Argo CD, in order to generate Applications for them.
+
+## Repository Host Generator
+
+The RepoHost generator uses the API of an SCMaaS provider to discover repositories. This fits well with many repos following the same GitOps layout patterns such as microservices.
+
+### Github
+
+The Github mode uses the Github API to scan and organization in either github.com or Github Enterprise.
+
+```yaml
+apiVersion: argoproj.io/v1alpha1
+kind: ApplicationSet
+metadata:
+  name: myapps
+spec:
+  generators:
+  - repoHost:
+      github:
+        # The Github organization to scan.
+        organization: myorg
+        # For Github Enterprise:
+        api: https://git.example.com/
+        # Reference to a Secret containing an access token. (optional)
+        tokenRef:
+          name: github-token
+          key: token
+  template:
+  # ...
+```
+
+* `organization`: Required name of the Github organization to scan. If you have multiple orgs, use multiple generators.
+* `api`: If using Github Enterprise, the URL to access it.
+* `tokenRef`: A Secret name and key containing the Github access token to use for requests. If not specified, will make anonymous requests which have a lower rate limit.
+
+For label filtering, the repository topics are used.
+
+### Filters
+
+Filters allow selecting which repositories to generate for. Filters are additive, specifying none will template every repository and each filter added will pare that down.
+
```

Comment:
This wouldn't be correct, it has to match every rule rather than at least one.

Reply:
Gotcha, my original second sentence was incorrect, but that got my thinking about how I would expect users to approach this design.

TLDR: I am proposing changing the outer filters to be processed as a logical OR list (in order to be included a repository must match at least one filter), and maintaining the conditions within a filter as an AND list (eg all must match)

Mainly I am making this proposal due to how the current behaviour handles scenario 3, below.

#### Scenario one: the user specifies no filters

```yaml
generators:
  repoHost:
    github:
      organization: hi
      # (...)
```

My guess at user expected behaviour: none of the repositories in the organization are filtered out.
Actual behaviour: Matches the user's expected behaviour.



#### Scenario two: the user specifies one filter

```yaml
generators:
  repoHost:
    github:
      organization: hi
      filters:
        - repositoryMatch: my-repos*
```

User expected behaviour: only repositories that match that one filter are included
Actual behaviour: Same.



#### Scenario three: the user specifies multiple filters

```yaml
generators:
  repoHost:
    github:
      organization: hi
      filters:
        - repositoryMatch: my-repo1
        - repositoryMatch: my-repo2
```

Expected: Both repository `my-repo1` and `my-repo` will be included.
Actual: no repositories will be included (unless, if I am not mistaken, the user happens to have a repository named `my-repo1-my-repo2` :smile:, since the PR looks for the regexp to match a substring)


Now, a response to this might be to say that the user is doing it wrong, and that they should be using reg-exps for the above case. But my expectation is that this behaviour will trip up many more users than it will 'click' with, and that familiarity with regular expression construction is a less common skill than one might think.

#### Proposal

So my proposal would be to switch the outer filters to be processed via logic OR, while maintaining logical AND for the inner conditions.  This would mean that scenario 3 above would match (my guess at) the user's expected behaviour.

What do you think? Is there some other user scenario that becomes much more difficult to express with this approach? It's definitely a balancing act between guessing at user expectations, trying to maximize the expressive power of the filter, and keeping the schema as simple/straightforward as possible/necessary.

Reply:
So the only reason I made it an array in the first place was specifically the `PathExists` test which takes just one path, so you would need multiple blocks to check more than one file and in that context ANDing made more sense. I agree it's confusing in aggregate though. I can change that to `PathsExist` which takes a `[]string` and then make the overall multi-filter be OR, or just remove multi-filter?

Reply:
Also another thing to factor in, you can get an OR behavior now with multiple generators but that will make redundant API calls which is probably not great because rate limits are a thing.

Reply:
I think that multi-filter with OR is good behavior for now. We don't want to make multiple calls with multiple generators

Reply:
I would say yes to multi-filter with OR... as for PathExists, I'm fine with a string array (the user MAY be able to use a  regex to accomplish the same thing with a non-string-array PathExists, but that is probably just complicating things for them) 

Re: OR behaviour with multiple generators, that's true. My expectation is it will be pretty common for folks to want to target multiple repository patterns in the list of repositories to target, so we may as well keep it in at the generator level to make things easier for them, as well as reducing repitition. It's true that it doesn't necessarily increase the expressive power of the generator, but it does improve the ergonomics (and does make it easier to avoid hitting API rate limits, as you said).

And thanks for walking us through your thinking, and helping us better understand the problem space :+1:, I tend to find schema design is naturally a lot of cognitive trial and error, as we try ideas on for size and think through how this affects various user scenarios. I'm really looking forward to getting this in (let me know if I can help!) :smile: .

Reply:
Regex on path checks unfortunately isn't easy with the way the github API works so I made it only take single paths to check (it's probing them directly). Sounds good on the rest though, will reorg things a bit.